### PR TITLE
Add RFC 91 text: GDALDataset::Close() method

### PR DESCRIPTION
**Pull request by rouault**
Rendered view in https://github.com/rouault/gdal/blob/rfc91_text/doc/source/development/rfc/rfc91_dataset_close.rst

### CODE REVIEW

**Comment by jjimenezshaw** [comment]
Thanks @rouault . I was one of those that used the ugly hacky error detection.

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+It modifies the GDALDataset::FlushCache() method and GDALFlushCache()
+function to return an error code, adds a GDALDataset::Close() virtual method and
+modifies GDALClose() to return an error code.
+
+Motivation
+----------
+
+A lot of drivers that have write capabilities may do write operations in their
+destructors, or just close the file descriptor owned by the dataset. Currently
+there is no clean way of catching errors that might occur. This recently bite
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:30]: bite -> bit

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+do something like the following:
+
+.. code-block:: c++
+
+    CPLErr eErrCodeBefore = CPLGetLastErrorType();
+    GDALClose(hOutDS);
+    if( eErrCodeBefore == CE_None && CPLGetLastErrorType() != CE_None )
+        printf("Error occurred during GDALClose()!\n")
+
+
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:45]: Remove extra empty line

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+This is not pretty and there is always the slight risk that something in the
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:46]: suggestions: 
not pretty -> awkward
there is always the slight risk -> there is a risk

No need to qualify the risk (at least to me) as peoples' idea of level of risks might be different.

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+     * something goes wrong, in particular in creation / update scenarios where
+     * file write or network communication might occur when finalizing the dataset.
+     *
+     * Implementations should be robust to this method to be called several times
+     * (on subsequent calls, it should do nothing and return CE_None).
+     * Once it has been called, no other method than Close() or the dataset
+     * destructor should be called. RasterBand or OGRLayer owned by the dataset
+     * should be assumed as no longer being valid.
+     *
+     * If a driver implements this method, it should also call it from its
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:76]: Don't give wiggle room: it should also call -> it must also call

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+Out-of-tree drivers that have write capabilities are also encouraged to
+implement GDALDataset::Close().
+
+Limitations
+-----------
+
+Not all drivers will be modified to implement Close() in the candidate
+implementation, and even those modified might call internal methods of the
+driver that don't do error propagation. Consequently further work might be
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:141]: Suggest: drop contraction and add comma

driver that do not do error propagation. Consequently, further work might be

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+needed on a case-by-case to improve driver implementations.
+
+SWIG Bindings
+-------------
+
+The destructor of gdal.Dataset is modified to test the return value of GDALClose()
+and emits a CPLError(CE_Failure, ...) if the error state is clean (normally
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:148]: is clean -> is not clean ??

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+The existing tests of the C/C++ command line utilities test GDALClose() modified
+behavior.
+
+Issues / pull requests
+----------------------
+
+Addresses https://github.com/OSGeo/gdal/issues/6886
+
+https://github.com/OSGeo/gdal/compare/master...rouault:gdal:dataset_FlushCache_return_CPLErr?expand=1 contains
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:172]: move contains to the following line

File: doc/source/development/rfc/rfc91_dataset_close.rst [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+a candidate implementation.
+
+It updates all drivers that implements GDALDataset::FlushCache().
```
Comment (schwehr) [inline doc/source/development/rfc/rfc91_dataset_close.rst:175]: Convert this paragraph to a list like this:

The candidate implementation does the following:

* implements GDALDataset::Close() for all drivers that derive from RawDataset.
* ...

**Review by schwehr — COMMENTED** [review]
Looks great to me!  Some minor suggestions that are just polish.

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 1 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 2 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 3 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 4 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 5 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 7 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 8 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst]: done

File: doc/source/development/rfc/rfc91_dataset_close.rst (hunk 6 above)
Comment (rouault) [reply doc/source/development/rfc/rfc91_dataset_close.rst:147]: no, the current formulation is intended. See commit https://github.com/rouault/gdal/commit/c3c37ea9359af4cf93ec00df9058e63b6096213f
This is to make sure that the gdal.UseExceptions() mode throws an exception. It needs CPLGetLastErrorType() == CE_Failure, so if GDALClose() != CE_None and CPLGetLastErrorType() == CE_None, we issue a CPLError(CE_Failure, ...)

**Comment by mloskot** [comment]
It looks good to me.

A minor addition to the example to push it to safer side

```cpp
MyDataset::~MyDataset()
{
   try {
      MyDataset::Close();
   } catch (...) {
      // swallow error, optionally log it but ensure the action never throws
   }
}
```

**Review by mloskot — APPROVED** [review]

**Comment by rouault** [comment]
> A minor addition to the example to push it to safer side

added

**Pull request merged**